src/java/org/apache/cassandra/schema/PartitionDenylist.java

```diff
@@ -0,0 +1,462 @@
+     */
+    private static class DenylistEntry
+    {
+        public final ImmutableSet<ByteBuffer> keys;
+        public final ImmutableSortedSet<Token> tokens;
+
+        public DenylistEntry()
+        {
+            keys = ImmutableSet.of();
+            tokens = ImmutableSortedSet.of();
+        }
+
+        public DenylistEntry(final ImmutableSet<ByteBuffer> keys, final ImmutableSortedSet<Token> tokens)
+        {
+            this.keys = keys;
+            this.tokens = tokens;
+        }
+    }
+
+    // synchronized on this
+    private int loadAttempts = 0;
+    private int loadSuccesses = 0;
+
+    public synchronized int getLoadAttempts()
+    {
+        return loadAttempts;
+    }
+    public synchronized int getLoadSuccesses()
+    {
+        return loadSuccesses;
+    }
+
+    /**
+     * Performs initial load of the partition denylist.  Should be called at startup and only loads if the operation
+     * is expected to succeed.  If it is not possible to load at call time, a timer is set to retry.
+     */
+    public void initialLoad()
+    {
+        if (!DatabaseDescriptor.enablePartitionDenylist())
+            return;
+
+        synchronized (this)
+        {
+            loadAttempts++;
+        }
+
+        // Check if there are sufficient nodes to attempt reading all the denylist partitions before issuing the query.
+        // The pre-check prevents definite range-slice unavailables being marked and triggering an alert.  Nodes may still change
+        // state between the check and the query, but it should significantly reduce the alert volume.
+        String retryReason = "Insufficient nodes";
+        try
+        {
+            if (readAllHasSufficientNodes() && load())
+            {
+                return;
+            }
+        }
+        catch (Throwable tr)
+        {
+            logger.error("Failed to load partition denylist", tr);
+            retryReason = "Exception";
+        }
+
+        // This path will also be taken on other failures other than UnavailableException,
+        // but seems like a good idea to retry anyway.
+        int retryInSeconds = DatabaseDescriptor.getDenylistInitialLoadRetrySeconds();
+        logger.info(retryReason + " while loading partition denylist cache.  Scheduled retry in {} seconds.",
+                    retryInSeconds);
+        ScheduledExecutors.optionalTasks.schedule(this::initialLoad, retryInSeconds, TimeUnit.SECONDS);
+    }
+
+    private boolean readAllHasSufficientNodes()
+    {
+        return RangeCommands.sufficientLiveNodesForSelectStar(SystemDistributedKeyspace.PartitionDenylistTable,
+                                                              DatabaseDescriptor.getDenylistConsistencyLevel());
+    }
+
+    /** Helper method as we need to both build cache on initial init but also on reload of cache contents and params */
+    private LoadingCache<TableId, DenylistEntry> buildEmptyCache()
+    {
+        return CacheBuilder.newBuilder()
```

Comment:
I've just mentioned. Guava's cache is deprecated in favor of Caffeine (they basically extracted the cache to a separate library). Currently all (at least most of) caches have been migrated to Caffeine. Please, do the same. The API is generally the same (or almost the same), probably just need to update imports.

Reply:
A touch different as caffeine takes an executor and has a different signature on asyncReload vs. reload w/async hacked in like we had, but clear enough.